prometheus/wrap.go

```diff
@@ -0,0 +1,172 @@
+// should not be used to add fixed labels to all metrics exposed.
+//
+// The Collector example demonstrates a use of WrapWith.
+func WrapWith(labels Labels, reg Registerer) Registerer {
+	return &wrappingRegisterer{
+		wrappedRegisterer: reg,
+		labels:            labels,
+	}
+}
+
+// WrapWithPrefix returns a Registerer wrapping the provided Registerer. Collectors
+// registered with the returned Registerer will be registered with the wrapped
+// Registerer in a modified way. The modified Collector adds the provided prefix
+// to the name of all Metrics it collects.
+//
+// WrapWithPrefix is useful to have one place to prefix all metrics of a
+// sub-system. To make this work, register metrics of the sub-system with the
+// wrapping Registerer returned by WrapWithPrefix.
+func WrapWithPrefix(prefix string, reg Registerer) Registerer {
+	return &wrappingRegisterer{
+		wrappedRegisterer: reg,
+		prefix:            prefix,
+	}
+}
+
+type wrappingRegisterer struct {
+	wrappedRegisterer Registerer
+	prefix            string
+	labels            Labels
+}
+
+func (r *wrappingRegisterer) Register(c Collector) error {
+	return r.wrappedRegisterer.Register(&wrappingCollector{
+		wrappedCollector: c,
+		prefix:           r.prefix,
+		labels:           r.labels,
+	})
+}
+
+func (r *wrappingRegisterer) MustRegister(cs ...Collector) {
+	for _, c := range cs {
+		if err := r.Register(c); err != nil {
+			panic(err)
+		}
+	}
+}
+
+func (r *wrappingRegisterer) Unregister(c Collector) bool {
+	return r.wrappedRegisterer.Unregister(&wrappingCollector{
+		wrappedCollector: c,
+		prefix:           r.prefix,
+		labels:           r.labels,
+	})
+}
+
+type wrappingCollector struct {
+	wrappedCollector Collector
+	prefix           string
+	labels           Labels
+}
+
+func (c *wrappingCollector) Collect(ch chan<- Metric) {
+	wrappedCh := make(chan Metric)
+	go func() {
+		c.wrappedCollector.Collect(wrappedCh)
+		close(wrappedCh)
+	}()
+	for m := range wrappedCh {
+		ch <- &wrappingMetric{
+			wrappedMetric: m,
+			prefix:        c.prefix,
+			labels:        c.labels,
+		}
+	}
+}
+
+func (c *wrappingCollector) Describe(ch chan<- *Desc) {
+	wrappedCh := make(chan *Desc)
+	go func() {
+		c.wrappedCollector.Describe(wrappedCh)
+		close(wrappedCh)
+	}()
+	for desc := range wrappedCh {
+		ch <- wrapDesc(desc, c.prefix, c.labels)
+	}
+}
+
+type wrappingMetric struct {
+	wrappedMetric Metric
+	prefix        string
+	labels        Labels
+}
+
+func (m *wrappingMetric) Desc() *Desc {
+	return wrapDesc(m.wrappedMetric.Desc(), m.prefix, m.labels)
+}
+
+func (m *wrappingMetric) Write(out *dto.Metric) error {
+	if err := m.wrappedMetric.Write(out); err != nil {
+		return err
+	}
+	if len(m.labels) == 0 {
+		// No wrapping labels.
+		return nil
+	}
+	for ln, lv := range m.labels {
+		out.Label = append(out.Label, &dto.LabelPair{
+			Name:  proto.String(ln),
+			Value: proto.String(lv),
+		})
+	}
+	sort.Sort(labelPairSorter(out.Label))
```

Comment:
I realize that we recommend this, but above all we want a stable label order, no? I wonder whether this might be over doing it and potentially causing problems in performance in rendering of the metrics output.

Reply:
Given how relatively infrequent scrapes are in comparison to other actions done by the client, I would guess that over all this has a relatively small effect on performance.

Reply:
The problem I'm seeing is that we sort potentially multiple times per sample as we sort at every stage of "unwrapping", I don't see the value add of sorting, when we already have a stable order. I realize the stable order may not apply for custom `prometheus.Collector` structs, but that should be part of the interface not the infrastructure enforcing it.

Reply:
Traditionally, client_golang has always provided "perfectly" sorted output, despite not being required by the exposition format spec. I would like to keep it that way for now to avoid surprises. In this concrete examples, the allocation of the LabelPair proto message will probably need more resources than the sorting. When we move to a faster protobuf library or to no protobuf at all, we can think about performance tradeoffs of sorting. But that's again a v0.10 issue to be discussed later.